README.md
# NAME

Text::MarkdownTable - Write Markdown syntax tables from data

# STATUS

[![Build Status](https://travis-ci.org/nichtich/Text-MarkdownTable.png)](https://travis-ci.org/nichtich/Text-MarkdownTable)
[![Coverage Status](https://coveralls.io/repos/nichtich/Text-MarkdownTable/badge.png)](https://coveralls.io/r/nichtich/Text-MarkdownTable)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/Text-MarkdownTable.png)](http://cpants.cpanauthors.org/dist/Text-MarkdownTable)

# SYNOPSIS

    my $table = Text::MarkdownTable->new;
    $table->add({one=>"a",two=>"table"});
    $table->add({one=>"is",two=>"nice"});
    $table->done;

    | one | two   |
    |-----|-------|
    | a   | table |
    | is  | nice  |

    Text::MarkdownTable->new( columns => ['X','Y','Z'], edges => 0 )
      ->add({a=>1,b=>2,c=>3})->done;

    X | Y | Z
    --|---|--
    1 | 2 | 3
    

# DESCRIPTION

This module can be used to write data in tabular form, formatted in
MultiMarkdown syntax. The resulting format can be used for instance to display
CSV data or to include data tables in Markdown files. Newlines and vertical
bars in table cells are replaced by a space character and cell values can be
truncated.

# CONFIGURATION

- file

    Filename, GLOB, scalar reference or [IO::Handle](https://metacpan.org/pod/IO::Handle) to write to (default STDOUT).

- fields

    Array, hash reference, or comma-separated list of fields/columns.

- columns

    Column names. By default field names are used.

- widths

    Column widths. By default column widths are calculated automatically to the
    width of the widest value. With given widths, the table is directly be written
    without buffering and large table cell values are truncated.

- header

    Include header lines. Enabled by default.

- edges

    Include border before first column and after last column. Enabled by default.
    Note that single-column tables don't not look like tables if edges are
    disabled.

- condense

    Write table unbuffered in condense format:

        one|two
        ---|---
        a|table
        is|nice

    Note that single-column tables are don't look like tables on condense format.

# METHODS

- add( $row )

    Add a row as hash reference. Returns the table instance.

- streaming

    Returns whether rows are directly written or buffered until `done` is called.

- done

    Finish and write the table unless it has already been written in `streaming`
    mode.

# SEE ALSO

See [Catmandu::Exporter::Table](https://metacpan.org/pod/Catmandu::Exporter::Table) for an application of this module that can be
used to easily convert data to Markdown tables.

Similar table-generating modules include:

- [Text::Table::Tiny](https://metacpan.org/pod/Text::Table::Tiny)
- [Text::TabularDisplay](https://metacpan.org/pod/Text::TabularDisplay)
- [Text::SimpleTable](https://metacpan.org/pod/Text::SimpleTable)
- [Text::Table](https://metacpan.org/pod/Text::Table)
- [Text::ANSITable](https://metacpan.org/pod/Text::ANSITable)
- [Text::ASCIITable](https://metacpan.org/pod/Text::ASCIITable)
- [Text::UnicodeBox::Table](https://metacpan.org/pod/Text::UnicodeBox::Table)
- [Table::Simple](https://metacpan.org/pod/Table::Simple)
- [Text::SimpleTable](https://metacpan.org/pod/Text::SimpleTable)
- [Text::SimpleTable::AutoWidth](https://metacpan.org/pod/Text::SimpleTable::AutoWidth)

# AUTHOR

Jakob Voß <jakob.voss@gbv.de>

# COPYRIGHT AND LICENSE

Copyright 2014- Jakob Voß

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
{{$NEXT}}

0.3.1  2015-07-08 12:12:59 CEST
    - fix minimal dependency on Moo
    - additional option 'edges' to disable first and last border
    - public method 'streaming'

0.3.0  2015-07-01 12:02:42 CEST
    - additional option 'header'

0.2.4  2014-10-23 00:17:08 CEST
    - fixed undefined $_ in simple.t test (Patrick Hochstenbach)

0.2.3  2014-07-16 09:45:11 CEST
    - improved performance by using printf
    - support undefined column names

0.2.2  2014-07-15 15:02:56 CEST
    - simplify option 'fh' with IO::Handle::Util

0.2.1  2014-07-11 10:14:33 CEST
    - implemented condense and unbuffered writing 

0.2.0   2014-07-10
    - inital release, based on Catmandu::Exporter::Table 0.1.2


lib/Text/MarkdownTable.pm
package Text::MarkdownTable;
use strict;
use warnings;
use 5.010;

our $VERSION = '0.3.1';

use Moo;
use IO::File;
use IO::Handle::Util ();

has file => (
    is      => 'ro',
    lazy    => 1,
    default => sub { \*STDOUT },
);

has fh => (
    is      => 'ro',
    lazy    => 1,
    default => sub {
        my $fh = $_[0]->file;
        $fh = ref $fh 
            ? IO::Handle::Util::io_from_ref($fh) : IO::File->new($fh,"w");
        die "invalid option file" if !$fh;
        binmode $fh, $_[0]->encoding;
        $fh;
    }
);

has encoding => (
    is      => 'ro',
    default => sub { ':utf8' }
);

has fields => (
    is     => 'rw',
    trigger => 1,
);

# TODO: ensure that number of columns is number of fields
has columns => (
    is      => 'lazy',
    coerce  => \&_coerce_list,
    builder => sub { $_[0]->fields }
);

has widths => (
    is      => 'lazy',
    coerce  => \&_coerce_list,
    builder => sub {
        $_[0]->_fixed_width(0);
        return [map { defined($_) ? length $_ : 0 } @{$_[0]->columns}]
    },
);

has header => (
    is => 'rw', 
    default => sub { 1 }
);

has edges => (
    is => 'rw',
    default => sub { 1 },
);

has condense => (
    is => 'rw',
);

has streaming => (is => 'rwp');

has _fixed_width => (is => 'rw', default => sub { 1 });

# TODO: duplicated in Catmandu::Exporter::CSV fields-coerce
sub _coerce_list {
    if (ref $_[0]) {
        return $_[0] if ref $_[0] eq 'ARRAY';
        return [sort keys %{$_[0]}] if ref $_[0] eq 'HASH';
    }    
    return [split ',', $_[0]];
}

sub _trigger_fields {
    my ($self, $fields) = @_;
    $self->{fields} = _coerce_list($fields);
    if (ref $fields and ref $fields eq 'HASH') {
        $self->{columns} = [ map { $fields->{$_} // $_ } @{$self->{fields}} ];
    }
}

sub add {
    my ($self, $data) = @_;
    unless ($self->fields) {
        $self->{fields} = [ sort keys %$data ]
    }
    my $fields = $self->fields;
    my $widths = $self->widths; # may set 
    my $row = [ ];

    if (!$self->streaming and ($self->condense or $self->_fixed_width)) {
        $self->_set_streaming(1);
        $self->_print_header if $self->header;
    }

    foreach my $col (0..(@$fields-1)) {
        my $field = $fields->[$col];
        my $width = $widths->[$col];

        my $value = $data->{$field} // "";
        $value =~ s/[\n|]/ /g;

        my $w = length $value;
        if ($self->_fixed_width) {
            if (!$width or $w > $width) {
                if ($width > 5) {
                    $value = substr($value, 0, $width-3) . '...';
                } else {
                    $value = substr($value, 0, $width);
                }
            }
        } else {
            $widths->[$col] = $w if !$width or $w > $width;
        }
        push @$row, $value;
    }

    $self->_add_row($row);
    $self;
}

sub _add_row {
    my ($self, $row) = @_;

    if ($self->streaming) {
        $self->_print_row($row);
    } else {
        push @{$self->{_rows}}, $row;
    }
}

sub done {
    my ($self) = @_;

    if ($self->{_rows}) {
        $self->_print_header if $self->header;
        $self->_print_row($_) for @{$self->{_rows}};
    }
}

sub _print_header {
    my ($self) = @_;
    my $fh     = $self->fh;

    $self->_print_row($self->columns);
    if ($self->condense) {
        $self->_print_row([ map { '-' x length $_ } @{$self->columns} ]);
    } elsif ($self->edges) {
        print $fh '|'.('-' x ($_+2)) for @{$self->widths};
        print $fh "|\n";
    } else {
        print $fh substr(join('|',map { '-' x ($_+2) } @{$self->widths}),1,-1);
        print $fh "\n";
    }
}

has _row_format => (
    is      => 'lazy',
    builder => sub {
        if ( $_[0]->condense ) {
            join("|",map {"%s"} @{$_[0]->fields})."\n"
        } elsif ( $_[0]->edges ) {
            join("",map {"| %-".$_."s "} @{$_[0]->widths})."|\n";
        } else {
            join(" | ",map {"%-".$_."s"} @{$_[0]->widths})."\n";
        }
    }
);

sub _print_row {
    my ($self, $row) = @_;
    printf {$self->fh} $self->_row_format, @{$row};
}

1;
__END__

=head1 NAME

Text::MarkdownTable - Write Markdown syntax tables from data

=begin markdown

# STATUS

[![Build Status](https://travis-ci.org/nichtich/Text-MarkdownTable.png)](https://travis-ci.org/nichtich/Text-MarkdownTable)
[![Coverage Status](https://coveralls.io/repos/nichtich/Text-MarkdownTable/badge.png)](https://coveralls.io/r/nichtich/Text-MarkdownTable)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/Text-MarkdownTable.png)](http://cpants.cpanauthors.org/dist/Text-MarkdownTable)

=end markdown

=head1 SYNOPSIS

  my $table = Text::MarkdownTable->new;
  $table->add({one=>"a",two=>"table"});
  $table->add({one=>"is",two=>"nice"});
  $table->done;

  | one | two   |
  |-----|-------|
  | a   | table |
  | is  | nice  |

  Text::MarkdownTable->new( columns => ['X','Y','Z'], edges => 0 )
  ->add({a=>1,b=>2,c=>3})->done;

  X | Y | Z
  --|---|--
  1 | 2 | 3
  
=head1 DESCRIPTION

This module can be used to write data in tabular form, formatted in
MultiMarkdown syntax. The resulting format can be used for instance to display
CSV data or to include data tables in Markdown files. Newlines and vertical
bars in table cells are replaced by a space character and cell values can be
truncated.

=head1 CONFIGURATION

=over

=item file

Filename, GLOB, scalar reference or L<IO::Handle> to write to (default STDOUT).

=item fields

Array, hash reference, or comma-separated list of fields/columns.

=item columns

Column names. By default field names are used.

=item widths

Column widths. By default column widths are calculated automatically to the
width of the widest value. With given widths, the table is directly be written
without buffering and large table cell values are truncated.

=item header

Include header lines. Enabled by default.

=item edges

Include border before first column and after last column. Enabled by default.
Note that single-column tables don't not look like tables if edges are
disabled.

=item condense

Write table unbuffered in condense format:

  one|two
  ---|---
  a|table
  is|nice

Note that single-column tables are don't look like tables on condense format.

=back

=head1 METHODS

=over

=item add( $row )

Add a row as hash reference. Returns the table instance.

=item streaming

Returns whether rows are directly written or buffered until C<done> is called.

=item done

Finish and write the table unless it has already been written in C<streaming>
mode.

=back

=head1 SEE ALSO

Mardown tables as implemented in this module are also known as "pipe tables".
See L<GitHub|https://help.github.com/articles/github-flavored-markdown/#tables>
and L<Pandoc|http://pandoc.org/README.html#extension-pipe_tables> documentation
of pipe tables for (informal) specifications.

See L<Catmandu::Exporter::Table> for an application of this module to to easily
convert data in many formats to Markdown tables.

Similar table-generating modules include:

=over

=item L<Text::Table::Tiny>

=item L<Text::TabularDisplay>

=item L<Text::SimpleTable>

=item L<Text::Table>

=item L<Text::ANSITable>

=item L<Text::ASCIITable>

=item L<Text::UnicodeBox::Table>

=item L<Table::Simple>

=item L<Text::SimpleTable>

=item L<Text::SimpleTable::AutoWidth>

=back

=encoding utf8

=head1 AUTHOR

Jakob Voß E<lt>jakob.voss@gbv.deE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2014- Jakob Voß

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Catmandu/Exporter/Table.pm
package Catmandu::Exporter::Table;

our $VERSION = '0.1.2';

use namespace::clean;
use Catmandu::Sane;
use Moo;

with 'Catmandu::Exporter';

has fields => (
    is     => 'rw',
    trigger => 1,
);

has header => (
    is      => 'lazy',
    coerce  => \&_coerce_list,
    builder => sub { $_[0]->fields }
);

has widths => (
    is      => 'lazy',
    coerce  => \&_coerce_list,
    builder => sub {
        $_[0]->fixed_width(0);
        return [map { length $_ } @{$_[0]->header}]
    },
);

has fixed_width => (is => 'rw', default => sub { 1 });
has condense    => (is => 'ro');

# TODO: duplicated in Catmandu::Exporter::CSV fields-coerce
sub _coerce_list {
    if (ref $_[0]) {
        return $_[0] if ref $_[0] eq 'ARRAY';
        return [sort keys %{$_[0]}] if ref $_[0] eq 'HASH';
    }    
    return [split ',', $_[0]];
}

sub _trigger_fields {
    my ($self, $fields) = @_;
    $self->{fields} = _coerce_list($fields);
    if (ref $fields and ref $fields eq 'HASH') {
        $self->{header} = [ map { $fields->{$_} } @{$self->{fields}} ];
    }
}

sub add {
    my ($self, $data) = @_;
    unless ($self->fields) {
        $self->{fields} = [ sort keys %$data ]
    }
    my $fields = $self->fields;
    my $widths = $self->widths;
    my $row = [ ];

    foreach my $col (0..(@$fields-1)) {
        my $field = $fields->[$col];
        my $width = $widths->[$col];

        my $value = $data->{$field} // "";
        $value =~ s/[\n|]/ /g;

        my $w = length $value;
        if ($self->fixed_width) {
            if ($w > $width) {
                if ($width > 5) {
                    $value = substr($value, 0, $width-3) . '...';
                } else {
                    $value = substr($value, 0, $width);
                }
            }
        } else {
            $widths->[$col] = $w if $w > $width;
        }
        push @$row, $value;
    }

    push @{$self->{_rows}}, $row;
}

# TODO: support ugly Markdown table (no alignment)

sub commit {
    my ($self) = @_;

    my $fh      = $self->fh;
    my $widths  = $self->widths;
    my $header = $self->header;
#    my $line = '-' x (sum(@$widths) + 2*@$widths + 2) . "\n";
 
    $self->print_multimarkdown_row($self->header);
    map { print $fh "|".('-' x ($_+2))} @$widths;
    print $fh "|\n";

    $self->print_multimarkdown_row($_) for @{$self->{_rows}};
}

sub print_multimarkdown_row {
    my ($self, $row) = @_;
    my $fh     = $self->fh;
    my $widths = $self->widths;
    foreach my $col (0..(@$widths-1)) {
        printf $fh "| %-".$widths->[$col]."s ", $row->[$col];
    }
    print $fh "|\n";
}

=head1 NAME

Catmandu::Exporter::Table - ASCII/Markdown table exporter

=head1 SYNOPSIS

  echo '{"one":"my","two":"table"} {"one":"is","two":"nice"}]' | \ 
  catmandu convert JSON --multiline 1 to Table
  | one | two   |
  |-----|-------|
  | my  | table |
  | is  | nice  |

  catmandu convert CSV to Table --fields id,name --header ID,Name < sample.csv
  | ID | Name |
  |----|------|
  | 23 | foo  |
  | 42 | bar  |
  | 99 | doz  |

=head1 DESCRIPTION

This L<Catmandu::Exporter> exports data in tabular form, formatted in
MultiMarkdown syntax. The resulting format can be used for instance to display
CSV data or to include data tables in Markdown files. Newlines and vertical
bars in table cells are replaced by a space character and cell values can be
truncated.

=head1 CONFIGURATION

=over

=item fields

Array, hash reference, or comma-separated list of fields/columns.

=item header

Column names. By default field names are used.

=item widths

Column widths. By default column widths are calculated automatically to the
width of the widest value. With custom width, large values may be truncated.

=back

=head1 METHODS

See L<Catmandu::Exporter>, L<Catmandu::Addable>, L<Catmandu::Fixable>,
L<Catmandu::Counter>, and L<Catmandu::Logger> for a full list of methods.

=head1 SEE ALSO

L<Catmandu::Exporter::CSV>, L<Text::TabularDisplay>, L<Text::SimpleTable>, and
L<Text::Table>.

=cut

1;
